README.md


[![Greenkeeper badge](https://badges.greenkeeper.io/carpages/gemini-gallery.svg)](https://greenkeeper.io/)

CHANGELOG.md
# CHANGELOG

## v1.1.0

  - Update Handlebars to v4

## v1.0.4

  - Fix CommonJS module requires

## v1.0.3

  - Fix linting errors

## v1.0.2

  - Fix linting grunt task

## v1.0.1

  - Remove jshint in favour of eslint with custom ruleset
  - Bump travis-ci node version
  - Fix CommonJS require statement to require the correct gemini-loader module

## v1.0.0

  - Updated to handlebars v3.0.3
  - Added UMD support
  - Updated to gemini-modal v1.0.0
  - Updated to gemini-carousel v1.0.0


test/gemini-gallery.edge.js
/* eslint key-spacing: ["error", { "align": "colon" }] */
requirejs.config({
  baseUrl: '../',
  paths  : {
    'underscore'               : 'bower_components/underscore/underscore',
    'jquery'                   : 'bower_components/jquery/dist/jquery',
    'handlebars'               : 'bower_components/handlebars/handlebars.runtime',
    'jquery.boiler'            : 'bower_components/jquery-boiler/jquery.boiler',
    'gemini.support'           : 'bower_components/gemini-support/gemini.support',
    'gemini'                   : 'bower_components/gemini-loader/gemini',
    'gemini.modal'             : 'bower_components/gemini-modal/gemini.modal',
    'gemini.modal.templates'   : 'bower_components/gemini-modal/templates',
    'gemini.carousel'          : 'bower_components/gemini-carousel/gemini.carousel',
    'gemini.carousel.templates': 'bower_components/gemini-carousel/templates',
    'gemini.lazyload'          : 'bower_components/gemini-lazyload/gemini.lazyload',
    'gemini.respond'           : 'bower_components/gemini-respond/gemini.respond',
    'gemini.fold'              : 'bower_components/gemini-fold/gemini.fold',
    'gemini.gallery.templates' : 'templates'
  }
});

require([ 'gemini', 'gemini.gallery' ], function( G ) {
  G( '#js-gallery' ).gallery();
});


gruntfile.js
'use strict';

var compassImporter = require( 'compass-importer' );

module.exports = function( grunt ) {
  // Load all grunt tasks
  require( 'load-grunt-tasks' )( grunt );

  // Show elapsed time at the end
  require( 'time-grunt' )( grunt );

  // Project configuration.
  grunt.initConfig({
    // Metadata.
    pkg: grunt.file.readJSON( 'package.json' ),
    banner: '/*! <%= pkg.name %> - v<%= pkg.version %> - ' +
      '<%= grunt.template.today("yyyy-mm-dd") %>\n' +
      '<%= pkg.homepage ? "* " + pkg.homepage + "\\n" : "" %>' +
      '* Copyright (c) <%= grunt.template.today("yyyy") %> <%= pkg.author.name %>;' +
      ' Licensed MIT */\n',
    // Task configuration.
    handlebars: {
      compile: {
        options: {
          umd: true,
          namespace: 'Templates.Default.Gallery',
          processName: function( filename ) {
            return filename.split( '/' ).pop().replace( '.hbs', '' );
          }
        },
        files: [{
          src: 'templates/*.hbs',
          dest: 'templates.js'
        }]
      }
    },
    qunit: {
      all: {
        options: {
          timeout: 10000,
          urls: [ 'http://localhost:9000/test/<%= pkg.name %>.html' ],
          page: {
            viewportSize: { width: 10000, height: 10000 }
          }
        }
      }
    },
    eslint: {
      options: {
        configFile: '.eslintrc'
      },
      target: [ 'gemini.gallery.js' ]
    },
    connect: {
      server: {
        options: {
          hostname: '*',
          port: 9000
        }
      }
    },
    sass: {
      options: {
        importer: compassImporter,
        includePaths: [ 'bower_components' ]
      },
      dist: {
        options: {
          outputStyle: 'expanded'
        },
        files: {
          'test/gemini-gallery.test.css': 'test/gemini-gallery.test.scss'
        }
      }
    },
    'saucelabs-qunit': {
      all: {
        options: {
          urls: [ 'http://localhost:9000/test/<%= pkg.name %>.html' ],
          build: process.env.TRAVIS_JOB_ID,
          browsers: [
            // iOS
            {
              browserName: 'iphone',
              platform: 'OS X 10.9',
              version: '7.1'
            },
            {
              browserName: 'ipad',
              platform: 'OS X 10.9',
              version: '7.1'
            },
            // Android
            {
              browserName: 'android',
              platform: 'Linux',
              version: '4.3'
            },
            // OS X
            {
              browserName: 'safari',
              platform: 'OS X 10.9',
              version: '7'
            },
            {
              browserName: 'safari',
              platform: 'OS X 10.8',
              version: '6'
            },
            {
              browserName: 'firefox',
              platform: 'OS X 10.9',
              version: '28'
            },
            // Windows
            {
              browserName: 'internet explorer',
              platform: 'Windows 8.1',
              version: '11'
            },
            {
              browserName: 'internet explorer',
              platform: 'Windows 8',
              version: '10'
            },
            {
              browserName: 'internet explorer',
              platform: 'Windows 7',
              version: '11'
            },
            {
              browserName: 'internet explorer',
              platform: 'Windows 7',
              version: '10'
            },
            {
              browserName: 'internet explorer',
              platform: 'Windows 7',
              version: '9'
            },
            {
              browserName: 'internet explorer',
              platform: 'Windows 7',
              version: '8'
            },
            {
              browserName: 'firefox',
              platform: 'Windows 7',
              version: '29'
            },
            {
              browserName: 'chrome',
              platform: 'Windows 7',
              version: '34'
            },
            // Linux
            {
              browserName: 'firefox',
              platform: 'Linux',
              version: '29'
            }
          ]
        }
      }
    }
  });

  // Default task.
  grunt.registerTask( 'default', [ 'sass', 'eslint'/*, 'connect', 'qunit' */ ]);
  grunt.registerTask( 'ci', [ 'default'/*, 'saucelabs-qunit' */ ]);
};


templates.js
(function(factory) {

if (typeof define === 'function' && define.amd) {

define(['handlebars'], factory);

} else if (typeof exports === 'object') {

module.exports = factory(require('handlebars'));

} else {

factory(Handlebars);

}

}(function(Handlebars) {

this["Templates"] = this["Templates"] || {};
this["Templates"]["Default"] = this["Templates"]["Default"] || {};
this["Templates"]["Default"]["Gallery"] = this["Templates"]["Default"]["Gallery"] || {};

this["Templates"]["Default"]["Gallery"]["gallery"] = Handlebars.template({"1":function(container,depth0,helpers,partials,data) {
    var stack1, helper, alias1=depth0 != null ? depth0 : {};

  return "    <li>\n\n      <div class=\"fit\">\n        <img class=\"lazy clickable\"\n          data-original=\""
    + container.escapeExpression(((helper = (helper = helpers.src || (depth0 != null ? depth0.src : depth0)) != null ? helper : helpers.helperMissing),(typeof helper === "function" ? helper.call(alias1,{"name":"src","hash":{},"data":data}) : helper)))
    + "\"\n"
    + ((stack1 = helpers.each.call(alias1,(depth0 != null ? depth0.screens : depth0),{"name":"each","hash":{},"fn":container.program(2, data, 0),"inverse":container.noop,"data":data})) != null ? stack1 : "")
    + "          >\n      </div>\n\n    </li>\n";
},"2":function(container,depth0,helpers,partials,data) {
    var helper, alias1=depth0 != null ? depth0 : {}, alias2=helpers.helperMissing, alias3="function", alias4=container.escapeExpression;

  return "            data-"
    + alias4(((helper = (helper = helpers.screen || (depth0 != null ? depth0.screen : depth0)) != null ? helper : alias2),(typeof helper === alias3 ? helper.call(alias1,{"name":"screen","hash":{},"data":data}) : helper)))
    + "=\""
    + alias4(((helper = (helper = helpers.src || (depth0 != null ? depth0.src : depth0)) != null ? helper : alias2),(typeof helper === alias3 ? helper.call(alias1,{"name":"src","hash":{},"data":data}) : helper)))
    + "\"\n";
},"compiler":[7,">= 4.0.0"],"main":function(container,depth0,helpers,partials,data) {
    var stack1;

  return "<div class=\"js-gallery-carousel carousel\">\n  <ul class=\"carousel__list\">\n\n"
    + ((stack1 = helpers.each.call(depth0 != null ? depth0 : {},(depth0 != null ? depth0.imgs : depth0),{"name":"each","hash":{},"fn":container.program(1, data, 0),"inverse":container.noop,"data":data})) != null ? stack1 : "")
    + "\n  </ul>\n</div>\n";
},"useData":true});

this["Templates"]["Default"]["Gallery"]["modal"] = Handlebars.template({"compiler":[7,">= 4.0.0"],"main":function(container,depth0,helpers,partials,data) {
    return "<div class=\"js-modal  modal\">\n  <div class=\"js-modal__content  modal__content  full\"></div>\n  <div class=\"js-modal__close  modal__close\">Close</div>\n</div>\n";
},"useData":true});

return this["Templates"]["Default"]["Gallery"];

}));

jquery.gallery.js
/**
 * @fileoverview

A jQuery plugin to quickly build galleries using a modal and carousel.

### Notes
- The gallery is built using a list of images. See example.

 *
 * @namespace jquery.gallery
 * @copyright Carpages.ca 2014
 * @author Matt Rose <matt@mattrose.ca>
 *
 * @requires underscore
 * @requires jquery
 * @requires jquery.boiler
 * @requires jquery.modal
 * @requires jquery.carousel
 * @requires jquery.lazyload
 * @requires jquery.respond
 *
 * @prop {object} templates {@link jquery.gallery#templates}
 *
 * @example
  <html>
    <ul id="js-gallery">
      <li>
        <a href="http://www.placetim.com/200/200/">
          <img src="http://www.placetim.com/200/200/">
        </a>
      </li>
      <li>
        <a href="http://www.placetim.com/250/250/">
          <img src="http://www.placetim.com/250/250/">
        </a>
      </li>
    </ul>
  </html>
 *
 * @example
  $('#js-gallery').gallery();
 */

define([
  'jquery-loader',
  'underscore',
  'jquery.gallery.templates',
  'jquery.boiler',
  'jquery.modal',
  'jquery.carousel',
  'jquery.lazyload',
  'jquery.respond'
], function($, _, T){

  $.boiler('gallery', {
    defaults: {
      /**
       * Precompiled Handlebar templates to replace default. Expecting 'gallery'
       * and 'modal'
       * @name jquery.gallery#templates
       * @type object
       * @default {}
       */
      templates: {}
    },

    data: ['title', 'description'],

    init: function(){
      var plugin = this;

      //Extend the templates
      plugin.T = $.extend(T, plugin.settings.templates);

      //Grab images
      plugin.imgs = _.map(plugin.$el.find('> li > a'), function(anchor, i){
        var $anchor = $(anchor);

        $anchor.click(function(e){
          e.preventDefault();
          plugin.open.call(plugin, i+1);
        });

        return {
          src: $anchor.attr('href')
        };
      });

      //Create the modal
      plugin.modal = new $.Modal({
        templates: plugin.T,
        fixed: true,
        stopPropagation: 'img, .js-gallery-nav',
        content: plugin.T.gallery({
          title: plugin.settings.title,
          description: plugin.settings.description,
          imgs: plugin.imgs
        })
      });

      //Activate Carousel
      plugin.$carousel = plugin.modal.$modal.find('.js-gallery-carousel');
      plugin.$carousel
        .carousel({scrollDelay:500, scrollEventDelay:500, loop:true})
        .lazyload({
          images:'img.lazy',
          effect:'fadeIn',
          threshold: window.$window.width() * 0.8
        });

      //Key press
      window.$window.on('keydown', function(e){

        if( !plugin.modal.$wrapper.hasClass('is-active') ) return;

        if(e.keyCode==37){//left

          plugin.$carousel.carousel('previous');

        } else if(e.keyCode==39){//right

          plugin.$carousel.carousel('next');

        } else if(e.keyCode==27){//esc

          plugin.modal.close();

        }
      });

      //Go to next when you click on one of the images
      plugin.$carousel.on('click', 'img', function(){
        plugin.$carousel.carousel('next');
      });

      //Fit images according to the size of the screen
      plugin.$carousel._fitImages();

      $.respond.bind('resize', function(e, scrn){
        plugin.$carousel._fitImages();
      });

    },

    /**
     * Open the gallery to a specific item
     *
     * @method
     * @name jquery.gallery#open
     * @param {integer} page The page to open up the gallery to
    **/
    open: function(page){
      var plugin = this;

      plugin.modal.open();
      plugin.$carousel
        .carousel('gotoPage', page, false)
        .lazyload('update');
    },

    /**
     * Close the gallery
     *
     * @method
     * @name jquery.gallery#close
    **/
    close: function(){
      this.modal.close();
    }
  });

  // Return the jquery object
  // This way you don't need to require both jquery and the plugin
  return $;

});


gemini.gallery.js
/* global Templates */
/**
 * @fileoverview

A Gemini plugin to quickly build galleries using a modal and carousel.

### Notes
- Default template requires the 'fit' module from gemini-css
- The gallery is built using a list of images. See example.

 *
 * @namespace gemini.gallery
 * @copyright Carpages.ca 2014
 * @author Matt Rose <matt@mattrose.ca>
 *
 * @requires gemini
 * @requires gemini.modal
 * @requires gemini.carousel
 * @requires gemini.lazyload
 * @requires gemini.respond
 *
 * @prop {integer} scrollSpeed {@link gemini.carousel#scrollSpeed}
 * @prop {array} screens {@link gemini.carousel#screens}
 * @prop {object} templates {@link gemini.gallery#templates}
 *
 * @example
  <html>
    <ul id="js-gallery">
      <li>
        <a href="http://www.placetim.com/200/200/">
          <img src="http://www.placetim.com/200/200/">
        </a>
      </li>
      <li>
        <a href="http://www.placetim.com/250/250/">
          <img src="http://www.placetim.com/250/250/">
        </a>
      </li>
    </ul>
  </html>
 *
 * @example
  G('#js-gallery').gallery();
 */

( function( factory ) {
  if ( typeof define === 'function' && define.amd ) {
    // AMD. Register as an anonymous module.
    define([
      'gemini',
      'gemini.gallery.templates',
      'gemini.modal',
      'gemini.carousel',
      'gemini.lazyload',
      'gemini.respond'
    ], factory );
  } else if ( typeof exports === 'object' ) {
    // Node/CommonJS
    module.exports = factory(
      require( 'gemini-loader' ),
      require( './templates.js' ),
      require( 'gemini-modal' ),
      require( 'gemini-carousel' ),
      require( 'gemini-lazyload' ),
      require( 'gemini-respond' )
    );
  } else {
    // Browser globals
    factory( G, Templates.Default.Gallery );
  }
}( function( $, T ) {
  var _ = $._;

  $.boiler( 'gallery', {
    defaults: {
      /**
       * The speed that the carousel scrolls at in milliseconds.
       * @name gemini.carousel#scrollSpeed
       * @type Integer
       * @default 500
       */
      scrollSpeed: 500,

      /**
       * An array of screens to look for in the anchors data attributes to use
       * when lazy loading the images
       * @name gemini.carousel#screens
       * @type Array
       * @default []
       */
      screens: [],

      /**
       * Precompiled Handlebar templates to replace default. Expecting 'gallery'
       * and 'modal'
       * @name gemini.gallery#templates
       * @type object
       * @default {}
       */
      templates: {}
    },

    data: [ 'title', 'description' ],

    init: function() {
      var plugin = this;

      // Extend the templates
      plugin.T = $.extend( T, plugin.settings.templates );

      // Grab images
      plugin.imgs = _.map( plugin.$el.find( '> li > a' ), function( anchor, i ) {
        var $anchor = $( anchor );

        $anchor.click( function( e ) {
          e.preventDefault();
          plugin.open( i + 1 );
        });

        var screens = [];
        _.each( plugin.settings.screens, function( scn ) {
          if ( $anchor.data( scn )) {
            screens.push({
              screen: scn,
              src: $anchor.data( scn )
            });
          }
        });

        return {
          src: $anchor.attr( 'href' ),
          screens: screens
        };
      });

      // Create the modal
      plugin.modal = new $.Modal({
        templates: plugin.T,
        fixed: true,
        stopPropagation: 'img, .js-gallery-nav',
        content: plugin.T.gallery({
          title: plugin.settings.title,
          description: plugin.settings.description,
          imgs: plugin.imgs
        })
      });

      // Activate Carousel
      plugin.$carousel = plugin.modal.$content.find( '.js-gallery-carousel' );
      plugin.$carousel
        .carousel({ scrollSpeed: plugin.settings.scrollSpeed, loop: true })
        .lazyload({
          images: 'img.lazy',
          effect: 'fadeIn',
          threshold: window.$window.width() * 0.8
        });

      // Key press
      window.$window.on( 'keydown', function( e ) {
        if ( !plugin.modal.$modal.hasClass( 'is-active' )) {
          return;
        }

        if ( e.keyCode === 37 ) {
          // left
          plugin.$carousel.carousel( 'previous' );
        } else if ( e.keyCode === 39 ) {
          // right
          plugin.$carousel.carousel( 'next' );
        } else if ( e.keyCode === 27 ) {
          // esc
          plugin.modal.close();
        }
      });

      // Go to next when you click on one of the images
      plugin.$carousel.on( 'click', 'img', function() {
        plugin.$carousel.carousel( 'next' );
      });

      // Fit images according to the size of the screen
      plugin.$carousel.find( '.fit' )._fit();

      $.respond.bind( 'resize', function( e, scrn ) {
        plugin.$carousel.find( '.fit' )._fit();
      });
    },

    /**
     * Open the gallery to a specific item
     *
     * @method
     * @name gemini.gallery#open
     * @param {integer} page The page to open up the gallery to
    **/
    open: function( page ) {
      var plugin = this;

      plugin.modal.open();
      plugin.$carousel
        .carousel( 'gotoPage', page, false )
        .lazyload( 'update' );
    },

    /**
     * Close the gallery
     *
     * @method
     * @name gemini.gallery#close
    **/
    close: function() {
      this.modal.close();
    }
  });

  // Return the jquery object
  // This way you don't need to require both jquery and the plugin
  return $;
}));
